option to set primary attachment for item

abaevbog opened the pull request:
Tool -> Manage Attachments has a menu item to set an attachment as a primary file by adding a relation to the parent item with attachment's itemID as the object.
Primary attachment will be opened next time on click on the item. If the primary attachment is moved to trash, it is ignored. When a primary attachment is deleted or moved to another parent, the relation is deleted.
getAttachments() function sorts attachments, and if there is a primary attachment, it is moved to the top of the list.

Fixes: #355
Fixes: #1715

abaevbog commented on chrome/content/zotero/standalone/standalone.js:
```diff
@@ -286,10 +288,24 @@ const ZoteroStandalone = new function() {
+						document.getElementById("file-menuitem-primaryAttachment").label = "Unmark as primary attachment";
+					}
+					else {
+						document.getElementById("file-menuitem-primaryAttachment").label = "Mark as a primary attachment";
```
The `menu-edit` label lives in zotero>mozila>menubar.ftl. Since this is also a menu item, it is better to use menubar.ftl instead of zotero.ftl?

dstillman replied:
No, files in the Mozilla directory are from Firefox and are bundled via script.

dstillman reviewed (commented):
Oops, found some unsubmitted comments here

abaevbog commented:
Re-pushed! 

I wonder if there should be a more accessible way to get to this functionality? Tools > Manage attachments makes sense but, visually, should we add something to the row of itemTree that has a primary attachment to make it more apparent? 

Alternatively (or in addition), we could add another row to the mini-itemBox table in the attachment's itemPane, below the attachment preview. We already have a row "Indexed: Yes/No" with a button next to it. Maybe we add another row below it "Primary attachment: Yes/No" and some kind of button to toggle this primary attachment status?

abaevbog commented on chrome/content/zotero/xpcom/data/item.js:
```diff
@@ -3706,6 +3734,15 @@ Zotero.Item.prototype.getBestAttachment = Zotero.Promise.coroutine(function* ()
 	if (!this.isRegularItem()) {
 		throw ("getBestAttachment() can only be called on regular items");
 	}
+	// If there is a non-deleted primary attachment for this item, return it
```
That is a great point! I moved sorting of the primary attachment to the top into SQL of `getBestAttachments`. It's also a bit more concise this way.

abaevbog commented:
Awesome! 

I do think there should be is a better way to display/change which attachment is marked as primary because Tools > Manage Attachments needs to be somewhat discovered. This was started before the redesign, so there were not as many options. But now that we have a better itemPane, maybe it could be added as a checkbox of sorts in the attachments pane next to each `attachments-container` of the parent item? Or as another row of the mini-itembox table of the `attachment-box` of each attachment, below "Indexed: Yes/No"?

abaevbog commented:
For this to not break syncing, allowed predicates in [dataserver](https://github.com/zotero/dataserver/blob/172584aa38a17a7e6ecca51e2bbc25b00450fad9/model/Relations.inc.php#L33) would need to be updated to include `zotero:primaryAttachment`

abaevbog committed:
allow to mark a child attachment as primary

- added Set/Unset as a Primary Attachment menuitem to
attachment's context menu to add/remove primaryAttachmentPredicate
relations from the parent item with the attachment's key.
- if such relations exists on parent item, item.getBestAttachments()
will sort that item to the very top of the list
- primary attachment appears at the top in the itemTree
and will be opened when parent item is clicked
- if the primary attachment is moved to trash, it is ignored
- when a primary attachment is deleted or moved to another parent,
the relation is deleted

Fixes: zotero#355
Fixes: zotero#1715

abaevbog commented:
Re notes from https://github.com/zotero/zotero/issues/1715#issuecomment-2656154267:


> Let's rebase this. (I did, but I can't push due to Git LFS changes since the last commit. Not sure if there's a setting in your GitHub account that would fix that or it's just a fact of Git LFS life.)

I rebased and re-pushed. I am not sure, to be honest... I never did anything special with LFS settings. But if my older PRs are the only ones causing issues, maybe there is something off. Is there some kind of specific error being thrown? 

> We should definitely take this out of Manage Attachments and put it somewhere more reasonable. Maybe it just goes in the items-list context menu? That seems like the most natural place for this. And I don't think we need to worry too much about overfilling that menu — if you're right-clicking on an attachment, you're already in fairly deep.

Yeah, I moved that option to the context menu after "Change Parent", and it seems to make a good amount of sense. Ultimately, if you are trying to find this option, this is the first place you would go to. So that is probably the best option. 

> Primary attachment (both automatic and explicit) should maybe always sort first? Sort of weird that it doesn't, at least without something more drastic like bolding the row, which seems excessive.

I believe this is supposed to be happening already... Since `item.getBestAttachments` sorts the item with predicate to the top, primary attachment should be moved to the top row like this. Is that not what you are seeing?

https://github.com/user-attachments/assets/c1643531-6e27-4866-8cf5-596eb5e8eb64


> Maybe you shouldn't be able to set this on an attachment that's already the automatic choice? I'm not sure why it'd ever be necessary. (Obviously you could end up deleting other attachments such that the one that became the default was already set, but that's fine.)

On one hand, it makes sense because then setting an item as a primary attachment is essentially a noop. On the other hand, what if you are about to add new attachments but want to make sure that the currently-only PDF is the one that will always open in the future, regardless what other attachments will be added later? Currently, you can mark it as primary and be sure nothing will move it. But if we were to disallow placing the primary mark on the default primary attachment, some other PDF might end up getting opened. So you would have to remember to go back and set the primary PDF after adding more attachments, which is a bit of an extra step. Not sure how frequently this scenario would emerge though and if it's a real concern.

abaevbog committed:
synchronously sort attachments for itemTree

So that the primary attachment can always be the first
child row without having to await to query the database
when the parent row is expanded.

abaevbog commented:
> So instead of special handling of the primary attachment, we should probably move the best attachment to the top. The natural issue with that is getBestAttachment() talks to the DB and is async, while toggleOpenState in itemTree is synchronous and needs to be fast.

Here, I added some synchronous sorting of the attachments in `item.getAttachments()` so we can get the best attachment (or at least the best candidate for it) and move it to the top without querying the database. How does this look? I think having the attachment that will open in the reader as the first one makes a lot of sense.

schmidma commented:
A concrete use case for me is replacing preprints with published versions. I often create a Zotero item while a work is unpublished or nearly published, then attach the final PDF once it is released. Zotero continues opening the older preprint by default, even though the published PDF is now the version I primarily want to use.

Is there still interest in moving this forward? @abaevbog, do you expect to continue working on this PR? @dstillman and @AbeJellinek, does the overall direction still look right?

It looks like the latest revision addressed the context-menu and attachment-sorting feedback, but the branch now conflicts with `main`. Dataserver support for `zotero:primaryAttachment` also seems unresolved. Was there a decision on whether the relation should store the attachment key or the full item URI, and how the server should handle it?

I'm mainly trying to understand the current status. If it has stalled and nobody plans to pick it up, I might consider helping revive it.
